### ISSUE TITLE
Admins can log in via Eauth using IE 11

### ISSUE THREAD

**Issue by lauraGgit** [USDAForestService/fs-open-forest-platform#405]
As an admin, I would like to be able to log in via Eauth when using Internet Explorer 11.

-Current resolution 11/14/2018: FS admins install chrome/use chrome for OF admin tasks.

## Notes
In what seems to be just production, we get a log-in loop.

## Acceptance Criteria
- [x] I can use eauth with IE 11.  (An FS administrator can log in to the admin side of christmas tree through eauth)

## Tasks
- [x] replace custom CORS logic with https://expressjs.com/en/resources/middleware/cors.html, using `optionsSuccessStatus: 200 // some legacy browsers (IE11, various SmartTVs) choke on 204
`
- [x] Remove any `PIP` headers
- [x] verify fix on IE11
- [x] PO approved.

## Definition of Done
- [x] Code unit / e2e tested and comply with QWASP
- [x] Server actions captured by logs
- [x] No new security vulnerabilities
- [x] Internal team code review
- [x] Documentation / readme.md / CircleCI updated
- [x] Docker updated
- [x] Compare finished design with mockup
- [x] Usability tested

**Comment by lauraGgit** [comment]
verified that this is the case in VPN

**Comment by lauraGgit** [comment]
@mtlaney could you take a look at this?

I think its that somehow that in IE the localStorage object is not getting set when the user is redirected back to logged-in from the EAuth.

The logs indicate that an admin user object is being created, and likely sent back to the frontend. 

However, when the frontend then checks `/auth/user` to see if its authorized, we are receiving  `401`. Because of the redirects, I'm having a very hard time tracking the storage settings in IE.

In staging, both the `requestingUrl` and the user object are successfully being set when an admin logs in.

**Comment by mtlaney** [comment]
@lauraGgit I'll check it out

**Comment by lauraGgit** [comment]
@mtlaney any progress on this?

**Comment by lauraGgit** [comment]
sourcemap as false is now pushed to master, will go and test a bit more.

**Comment by lauraGgit** [comment]
even with the source maps as false this is still a login loop.

**Comment by adborden** [comment]
https://openforest.fs.usda.gov/

**Comment by adborden** [comment]
I looked into this with @lauraGgit today, now that source maps are enabled it's a little easier to debug, although it's  still very hard to understand what's going wrong. From what I can tell, after logging in through eauth, the callback to fs-intake-api is called, which should be setting the session cookie and redirecting to openforest.usda.gov. After redirecting, the front end requests the user object from the API. This is a CORS request, and should be send with credentials (with the session cookie). It's at this point where things go wrong. IE11* doesn't make the request with the session cookie so the API can't determine if the request is authorized and instead returns 401 (unauthorized). The app then redirects to eauth which recognizes the user and immediately redirects to the saml callback on the API, and the loop continues.

The best guess I have so far is that we're missing a [P3P] header. IE (only old versions of IE at that) is the only browse that implemented P3P and will block CORS requests if a P3P header is not present and doesn't match the default privacy policy. We can easily add a basic P3P header to see if that changes the behavior.

**Comment by adborden** [comment]
Adding the P3P header didn't have any effect.

I re-traced some steps with @lauraGgit . We confirmed that the session cookie on fs-intake-api is getting set. We can browse directly to `fs-intake-api.app.cloud.gov/auth/user`, see the cookie get sent, and we get back a valid JSON user.

We are still unsure why the CORS request is behaving differently on production vs staging. 

We are still unsure why this only repros in the Horizon VM. I've failed to reproduce the issue in other VMs (Win7 + IE11, Win8 + IE11). It's possible that there is something specifically configured by GSA IT or FS/USDA IT that triggers the behavior. If someone has IE11 in a non-government machine to test with, it would be good to know if this also reproduces in a native IE11.

We're currently looking closer at how the cookie is being set. I noticed that on staging in Chrome, I'm not seeing the cookie get set with the `domain` or `secure` attributes set, despite being configured in code. #453 and #455 will clean that up.

We've also reached out to a few other folks who have more IE experience.

**Comment by adborden** [comment]
I just pushed a tweak in #455 that updates where we send the P3P header. This might do it.

**Comment by lauraGgit** [comment]
@adborden Just verified that this is in fact working! 🎉 💯 .
@ASprinkle to verify this one, you will need to use the openforest.fs.usda.gov environment.

**Comment by ASprinkle** [comment]
I tried this again today before logging off: 
I was able to get here without issue: https://openforest.fs.usda.gov/christmas-trees/forests
When I tried this one https://openforest.fs.usda.gov/admin/christmas-trees/season-dates this is where it took me after a few different page flashes between eAuth, Blank Page and "logging in" Open Forest wait circle.
![image](https://user-images.githubusercontent.com/33435621/48451960-c0f62180-e761-11e8-8e43-d6b4ccec8d31.png)

**Comment by ASprinkle** [comment]
From Stand Up: "Won't Fix" label was for pilot, not as a long-term solution. Need to reach out to FS IT for long-term IE solutions for Open Forest.

**Comment by lauraGgit** [comment]
@ASprinkle if we change the url of the API then we can overwrite this concern

**Comment by lauraGgit** [comment]
- this is something with specifically FS's IE. Since this works correctly in staging, we surmise that it is a cross-domain issue. If the api url were changed to either openforest.fs.usda.gov/api - or routes like that or openforest-api.fs.usda.gov this would fix this bug.

**Comment by ASprinkle** [comment]
Thank you for the update on this @lauraGgit!  @davemcorwin perhaps we can discuss more during planning tomorrow morning?

**Comment by davemcorwin** [comment]
👍

**Comment by davemcorwin** [comment]
Previous attempts involved adding a `P3P` header to some of our requests, some of which were removed in https://github.com/18F/fs-open-forest-platform/pull/622.

For our next attempt, I recommend removing the remaining reference in `middleware.es6` and replacing our custom CORS implementation with https://expressjs.com/en/resources/middleware/cors.html with the `"optionsSuccessStatus": 204` option to handle IE11.

This has the added benefit of not reinventing the wheel and relying on a well used and tested library for CORS. If this does not fix IE11 issues, then we can evaluate updating our infrastructure to utilizing more same-domain requests.

**Comment by ASprinkle** [comment]
Success accessing staging and admin sites for Christmas trees and Special Uses!!

While there is a screen shot of the PIV card authentication, I was not prompted to enter my pin.

here are screen shots of what I encountered when accessing Christmas tree admin:
![image](https://user-images.githubusercontent.com/33435621/59129303-52f19180-8921-11e9-98e2-c6b3645d0515.png)

![image](https://user-images.githubusercontent.com/33435621/59129310-584edc00-8921-11e9-860c-7f9eda58598b.png)

![image](https://user-images.githubusercontent.com/33435621/59129319-5e44bd00-8921-11e9-895c-1d467cb377d9.png)

![image](https://user-images.githubusercontent.com/33435621/59129325-61d84400-8921-11e9-95ec-16756cdbb8c1.png)

**Comment by davemcorwin** [comment]
@ASprinkle I'm not sure what you mean by "While there is a screen shot of the PIV card authentication, I was not prompted to enter my pin." Were you prompted to authenticate with that popup instead of EAuth??

**Comment by ASprinkle** [comment]
@davemcorwin I was just noting that I was routed through eAuth and didn't need to enter my pin.  I wasn't sure if that was intended to be that way or not, it is certainly convenient. I was prompted in staging to enter the staging user name and password. As far as my review on this is concerned, this is PO approved - assuming all the tasks have been completed above. Thanks!

**Comment by davemcorwin** [comment]
Okay, it was heard to tell what your experience was from the screenshots. The experience with IE11 should be no different than chrome. You may want to try explicitly logging out from EAuth to make sure you have to log back in. I was able to test on IE11 and it looked good, but always nice to confirm!

**Comment by mwreiss** [comment]
@ASprinkle @davemcorwin I tested on my FS laptop on Windows 10 IE 11, it looked good to me.

**Comment by ASprinkle** [comment]
@mwreiss @davemcorwin it's all working for me. I have had no issues accessing admin staging sites, or production staging sites from IE.

**Comment by briandavidson** [comment]
@mtlaney do you know if this can be closed at this point?

**Comment by ASprinkle** [comment]
@briandavidson @mtlaney moving to "Done" after "Ready for Master" is good. We'll move stories over to closed at the end of the sprint, or beginning of next sprint. 

I will continue to test stories after they are pushed to production.  In the instance of testing this story, I have not been able to log into Christmas tree Admin environment in production through IE.  I have successfully logged in to the Christmas tree Admin environment in staging through IE.  Can you take another look at the push on this, and see if there isn't something else going on? If it's more work, we'll move it back to in progress.

**Comment by ASprinkle** [comment]
I was able to access admin staging site through IE today.  I was prompted for my PIV pin, and was successfully routed to the correct page! 
![image](https://user-images.githubusercontent.com/33435621/60466553-cc229280-9c08-11e9-971a-e78c8b2bd1d0.png)

**Comment by carlsonem** [comment]
Need to update eAuth for updating production

**Comment by ASprinkle** [comment]
From Stand up 7/9/2019: Still waiting on the upgrade to NextGen eAuth before we can continue working on this story.

**Comment by aaronburk** [comment]
Need information from someone at eauth for ? we need to move to next gen in production. 

"should be resolved" in prod?  Need to be pushed  to production on August 15th.

**Comment by mwreiss** [comment]
OF Prod was upgraded to Next Gen eAuth successfully!  

 After testing: 
1) Microsoft Edge - on FS GFE/FS network successfully tested and redirects to the correct admin page
2) Microsoft Edge - on home system successfully tested and redirects to the correct admin page
3) Chrome -  on FS GFE/FS network successfully tested and redirects to the correct admin page
4) Chrome - on home system successfully tested and redirects to the correct admin page 
5) IE 11 - on FS GFE/FS network failed to redirect to the correct admin page (sent to "beta" page) (Possible policy enforced in the IE 11 browser that requires exception).
6) IE 11 - on home system successfully tested and redirects to the correct admin page

**Comment by mwreiss** [comment]
Next Gen eAuth:

![image](https://user-images.githubusercontent.com/50459930/63173865-66e2ff00-c006-11e9-9b54-bb63611acad8.png)

**Comment by carlsonem** [comment]
eAuth upgrade has been completed.
